### PR TITLE
Feat/issue 1698 global quiet

### PR DESCRIPTION
**Pull request by dharrigan**
Hi,

Here's a first pass at introducing `quiet recipes`. I'm happy to take unboard any thoughts you may have. Not quite sure about the name `set quiet-recipes`, but was struggling to come up with something better (maybe `set globally-quiet`?) :-)

Tests pass (with my additional ones added).

Thank you for your consideration.

#Fixes 1698 

-=david=-

### CODE REVIEW

**Comment by casey** [comment]
Thanks for the PR!

Some thoughts:

- Let's call the setting `quiet`, so you can do `set quiet`.
- Let's call the attribute `[noquiet]`
- We could also not have an attribute, and invert the meaning of `@` if `set quiet` is used. So if you `set quiet`, you can do `@recipe:`, which would make that recipe not quiet`. `@` is a bit confusing though, so I'm not convinced either way on this.

**Comment by dharrigan** [comment]
Hi,

Thank you for your helpful feedback. I'll make the changes within the next few days :-)

With regards the inversion of the `@`. I thought about this and I decided on the principle of least surprise (for the user). They are used to `@` to signify that the recipe (or line) should be quiet, so I didn't want to break that by introducing an inverted case. I would rather accrete changes than break expectations. I then thought that by introducing the attribute of `[no-quiet-recipe]` (now `[noquiet]`), they could then choose to make it explicit if they want an already demarked recipe of `@` to become non-quiet when the global attribute is used.

So, no surprises was my thoughts behind that one :-)

-=david=-

**Comment by dharrigan** [comment]
A question...

You've recommended the attribute to be renamed to `noquiet`. However, there is the precedent of `no-exit-message`, with a hyphen between the `no` and the action. Would it be better to be consisent and call the attribute `no-quiet`?

Thank you.

**Comment by casey** [comment]
Ah, yeah, good catch. In that case it should be `no-quiet`.

**Comment by dharrigan** [comment]
Hi!

I was wondering if you have had a chance to consider to merge in this PR? Anything else I can do?

Thank you.

-=david=-

File: README.md [hunk 1]
```diff
@@ -188,10 +188,10 @@ You can also set the shell using command-line arguments. For example, to use Pow
     <tr>
       <td><a href="https://www.gentoo.org">Gentoo Linux</a></td>
       <td><a href="https://wiki.gentoo.org/wiki/Portage">Portage</a></td>
-      <td><a href="https://github.com/gentoo-mirror/guru/tree/master/sys-devel/just">guru/sys-devel/just</a></td>
+      <td><a href="https://github.com/gentoo-mirror/dm9pZCAq/tree/master/sys-devel/just">dm9pZCAq/sys-devel/just</a></td>
```
Comment (casey) [inline README.md]: It looks like there are some unrelated changes, and merge conflicts in README.md.

File: README.md [hunk 2]
```diff
@@ -1243,21 +1244,20 @@ These functions can fail, for example if a path does not have an extension, whic
...
-| `[no-cd]`<sup>1.9.0</sup>           | Don't change directory before executing recipe. |
-| `[no-exit-message]`<sup>1.7.0</sup> | Don't print an error message if recipe fails.   |
-| `[private]`<sup>1.10.0</sup>        | See [Private Recipes](#private-recipes).        |
-| `[unix]`<sup>1.8.0</sup>            | Enable recipe on Unixes. (Includes MacOS).      |
-| `[windows]`<sup>1.8.0</sup>         | Enable recipe on Windows.                       |
+| Name                                | Description                                                    |
+| ----------------------------------- | -----------------------------------------------                |
+| `[no-cd]`<sup>1.9.0</sup>           | Don't change directory before executing recipe.                |
+| `[no-exit-message]`<sup>1.7.0</sup> | Don't print an error message if recipe fails.                  |
+| `[no-quiet]`<sup>?</sup>            | Override globally quiet recipes and always echo out the recipe |
```
Comment (casey) [inline README.md]: We should also add `[quiet]`.

File: README.md [hunk 3]
```diff
@@ -2221,6 +2220,33 @@ goodbye
 # all done!
 ```
 
+In addition, an entire Justfile can be marked as being globally quiet by
+setting `set quiet` in the Justfile. For example:
```
Comment (casey) [inline README.md]:
```suggestion
All recipes in a Justfile can be made quiet with `set quiet`:
```

File: README.md [hunk 4]
```diff
@@ -2221,6 +2220,33 @@ goodbye
...
+
+foo:
+  echo "This is quiet"
+
+@foo2:
+  echo "This is also quiet"
+```
+
+If the recipe has the attribute of `[no-quiet]`, then the setting is
+ignored: For example:
```
Comment (casey) [inline README.md]:
```suggestion
If a recipe has`[no-quiet]`, then the setting is ignored:
```

File: README.md [hunk 5]
```diff
@@ -2338,6 +2364,7 @@ import 'foo/bar.just'
 
 a: b
   @echo A
+
```
Comment (casey) [inline README.md]:
```suggestion
```

File: src/attribute.rs [hunk 6]
```diff
@@ -12,6 +12,7 @@ pub(crate) enum Attribute {
   NoCd,
   NoExitMessage,
   Private,
+  NoQuiet,
```
Comment (casey) [inline src/attribute.rs:15]: These enum variants should be sorted.

File: src/attribute.rs [hunk 7]
```diff
@@ -31,7 +32,12 @@ mod tests {
   use super::*;
 
   #[test]
-  fn to_str() {
+  fn no_exit_message_to_str() {
```
Comment (casey) [inline src/attribute.rs]:
```suggestion
  fn to_str() {
```

File: src/attribute.rs [hunk 8]
```diff
@@ -31,7 +32,12 @@ mod tests {
...
     assert_eq!(Attribute::NoExitMessage.to_str(), "no-exit-message");
   }
+
+  #[test]
```
Comment (casey) [inline src/attribute.rs]: We can remove this test. The other test is there just to make sure that we're serializing attribute names in kebab-case, but we don't need to test every variant.

File: src/recipe.rs [hunk 9]
```diff
@@ -103,6 +103,10 @@ impl<'src, D> Recipe<'src, D> {
     !self.attributes.contains(&Attribute::NoExitMessage)
   }
 
+  fn no_quiet(&self) -> bool {
+    !self.attributes.contains(&Attribute::NoQuiet)
```
Comment (casey) [inline src/recipe.rs]: Should this be inverted, i.e.:

```suggestion
    self.attributes.contains(&Attribute::NoQuiet)
```

File: src/recipe.rs [hunk 10]
```diff
@@ -150,7 +154,10 @@ impl<'src, D> Recipe<'src, D> {
       }
       let mut evaluated = String::new();
       let mut continued = false;
-      let quiet_command = lines.peek().map_or(false, |line| line.is_quiet());
+
+      let quiet_command = self.no_quiet() && context.settings.quiet
```
Comment (casey) [inline src/recipe.rs]: This is used to determine if we should strip a leading `@` from the line, so we should only do so if it starts with an `@`:

```suggestion
      let quiet_line = lines.peek().map_or(false, |line| line.is_quiet());
```

File: src/recipe.rs [hunk 11]
```diff
@@ -193,7 +200,9 @@ impl<'src, D> Recipe<'src, D> {
 
       if config.dry_run
         || config.verbosity.loquacious()
-        || !((quiet_command ^ self.quiet) || config.verbosity.quiet())
+        || !((quiet_command ^ self.quiet)
+          || (context.settings.quiet && self.no_quiet())
+          || config.verbosity.quiet())
```
Comment (casey) [inline src/recipe.rs]: I think this should be something like:

```suggestion
        || !((quiet_line ^ self.quiet) || (context.settings.quiet && !self.no_quiet()) || config.verbosity.quiet())
```

**Review by casey — CHANGES_REQUESTED** [review]
Sorry I was slow getting to this! Check out my review comments.

File: README.md (hunk 1 above)
Comment (dharrigan) [reply README.md]: Fixed

File: README.md (hunk 2 above)
Comment (dharrigan) [reply README.md]: There isn't a `[quiet]` attribute, just `[no-quiet]` which overrides the `set quiet` if that's been set.

File: README.md (hunk 3 above)
Comment (dharrigan) [reply README.md]: done

File: README.md (hunk 4 above)
Comment (dharrigan) [reply README.md]: done

File: README.md (hunk 5 above)
Comment (dharrigan) [reply README.md]: done

File: src/attribute.rs (hunk 7 above)
Comment (dharrigan) [reply src/attribute.rs]: done

File: src/attribute.rs (hunk 8 above)
Comment (dharrigan) [reply src/attribute.rs]: done

File: src/recipe.rs (hunk 9 above)
Comment (dharrigan) [reply src/recipe.rs]: done

File: src/recipe.rs (hunk 10 above)
Comment (dharrigan) [reply src/recipe.rs]: done

File: src/recipe.rs (hunk 11 above)
Comment (dharrigan) [reply src/recipe.rs]: done

**Comment by dharrigan** [comment]
Hi!

All changes asked for applied. When you can spare the time, have a looksee and see what you think :-)

Thank you.

-=david=-

**Review by casey — CHANGES_REQUESTED** [review]
Sorry for letting this sit! This looks good, it just needs integration tests, which can go in `tests/quiet.rs`. The tests in that file are using the `test!` macro, which I try to avoid these days, so for examples of how to write tests, check out `tests/modules.rs`. In particular, all the interactions with the quiet setting, the noquiet attribute, quiet lines, and the `@` prefix to both recipes in lines, should be tested.

**Comment by dharrigan** [comment]
Hi,

Additional integration tests done. I hope this is looking better.

-=david=-

**Review by casey — APPROVED** [review]
Looks good to me!

**Comment by casey** [comment]
Thanks for the PR! I tweaked it a little bit, simplifying the tests and removing an `if` statement that would never be executed.

This is easily the worst conditional I've ever seen, amazingly using `||`, `!`, `^`, and `&&`:

```rust
      if config.dry_run
        || config.verbosity.loquacious()
        || !((quiet_line ^ self.quiet)
          || (context.settings.quiet && !self.no_quiet())
          || config.verbosity.quiet())
      {
```

But so be it! Functionality for making things quiet or not quiet has just evolved organically over time, so I guess that's just how it is 😂

**Pull request merged**